broker-core/src/test/java/io/zeebe/broker/workflow/processor/variable/UpdateVariableDocumentProcessorTest.java

```diff
@@ -0,0 +1,205 @@
+  }
+
+  @After
+  public void tearDown() {
+    reset(variablesState, controller);
+  }
+
+  @Test
+  public void shouldRejectIfNoScopeFound() {
+    // given
+    final TypedRecord<VariableDocumentRecord> command = newCommand();
+
+    // when
+    processor.onCommand(command, controller);
+
+    // then
+    assertRejection(RejectionType.NOT_FOUND);
+  }
+
+  @Test
+  public void shouldRejectOnMsgpackReadError() {
+    // given
+    final ElementInstance instance = newElementInstance();
+    final TypedRecord<VariableDocumentRecord> command = newCommand();
+    command
+        .getValue()
+        .setScopeKey(instance.getKey())
+        .setUpdateSemantics(VariableDocumentUpdateSemantic.PROPAGATE)
+        .setDocument(new UnsafeBuffer());
+
+    // when
+    doThrow(MsgpackReaderException.class)
+        .when(variablesState)
+        .setVariablesFromDocument(anyLong(), any());
+    processor.onCommand(command, controller);
+
+    // then
+    assertRejection(RejectionType.INVALID_ARGUMENT);
+  }
+
+  @Test
+  public void shouldApplyPropagateUpdateOperation() {
+    // given
+    final DirectBuffer document = asMsgPack("a", 1);
+    final ElementInstance instance = newElementInstance();
+    final TypedRecord<VariableDocumentRecord> command = newCommand();
+    command
+        .getValue()
+        .setScopeKey(instance.getKey())
+        .setUpdateSemantics(VariableDocumentUpdateSemantic.PROPAGATE)
+        .setDocument(document);
+
+    // when
+    processor.onCommand(command, controller);
+
+    // then
+    assertAccepted(command);
+    verify(variablesState, times(1)).setVariablesFromDocument(instance.getKey(), document);
+  }
+
+  @Test
+  public void shouldApplyLocalUpdateOperation() {
+    // given
+    final DirectBuffer document = asMsgPack("a", 1);
+    final ElementInstance instance = newElementInstance();
+    final TypedRecord<VariableDocumentRecord> command = newCommand();
+    command
+        .getValue()
+        .setScopeKey(instance.getKey())
+        .setUpdateSemantics(VariableDocumentUpdateSemantic.LOCAL)
+        .setDocument(document);
+
+    // when
+    processor.onCommand(command, controller);
+
+    // then
+    assertAccepted(command);
+    verify(variablesState, times(1)).setVariablesLocalFromDocument(instance.getKey(), document);
+  }
+
+  @Test
+  public void shouldPublishUpdatedRecord() {
+    // given
+    final DirectBuffer document = asMsgPack("a", 1);
+    final ElementInstance instance = newElementInstance();
+    final TypedRecord<VariableDocumentRecord> command = newCommand();
+    command
+        .getValue()
+        .setScopeKey(instance.getKey())
+        .setUpdateSemantics(VariableDocumentUpdateSemantic.PROPAGATE)
+        .setDocument(document);
+
+    // when
+    processor.onCommand(command, controller);
+
+    // then
+    assertAccepted(command);
+    verify(variablesState, times(1)).setVariablesFromDocument(instance.getKey(), document);
```

Comment:
I think you missed here an assertion, that the updated record is written?

Reply:
It's easy to miss but it uses two different methods on the `variablesState`, one is `setVariablesFromDocument` and the other is `setVariablesLocalFromDocument`